File: assignment 10/python/programs/3.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def is_full_binary_tree(root):
    if not root:
        return True

    if not root.left and not root.right:
        return True

    if root.left and root.right:
        return is_full_binary_tree(root.left) and is_full_binary_tree(root.right)

    return False

def is_complete_binary_tree(root):
    if not root:
        return True

    queue = [root]
    is_last_level = False

    while queue:
        current_node = queue.pop(0)

        if current_node is None:
            is_last_level = True
        else:
            if is_last_level:
                return False

            queue.append(current_node.left)
            queue.append(current_node.right)

    return True

# Example usage:
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)
root.right.left = Node(6)

if is_full_binary_tree(root):
    print("The tree is a full binary tree.")
else:
    print("The tree is not a full binary tree.")

if is_complete_binary_tree(root):
    print("The tree is a complete binary tree.")
else:
    print("The tree is not a complete binary tree.")

File: assignment 10/c/programs/3.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Define the structure for a binary tree node
struct Node
{
    int data;
    struct Node *left;
    struct Node *right;
};

// Function to create a new node
struct Node *createNode(int data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Function to check if a binary tree is full
bool isFullTree(struct Node *root)
{
    if (root == NULL)
    {
        return true;
    }

    // If a node has either 0 or 2 children, continue checking
    if ((root->left && root->right) || (!root->left && !root->right))
    {
        return isFullTree(root->left) && isFullTree(root->right);
    }
    else
    {
        return false;
    }
}

// Function to check if a binary tree is complete
bool isCompleteTree(struct Node *root, int index, int nodeCount)
{
    if (root == NULL)
    {
        return true;
    }

    if (index >= nodeCount)
    {
        return false;
    }

    return isCompleteTree(root->left, 2 * index + 1, nodeCount) &&
           isCompleteTree(root->right, 2 * index + 2, nodeCount);
}

int countNodes(struct Node *root)
{
    if (root == NULL)
    {
        return 0;
    }

    return 1 + countNodes(root->left) + countNodes(root->right);
}

int main()
{
    struct Node *root = createNode(1);
    root->left = createNode(2);
    root->right = createNode(3);
    root->left->left = createNode(4);
    root->left->right = createNode(5);

    int nodeCount = countNodes(root);

    if (isFullTree(root))
    {
        printf("The tree is a full binary tree.\n");
    }
    else
    {
        printf("The tree is not a full binary tree.\n");
    }

    if (isCompleteTree(root, 0, nodeCount))
    {
        printf("The tree is a complete binary tree.\n");
    }
    else
    {
        printf("The tree is not a complete binary tree.\n");
    }

    return 0;
}
